nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java

```diff
@@ -0,0 +1,342 @@
+        }
+    };
+
+    public static final Validator BrokerValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            try{
+                URI brokerURI = new URI(input);
+                if (!"".equals(brokerURI.getPath())) {
+                    return new ValidationResult.Builder().subject(subject).valid(false).explanation("the broker URI cannot have a path. It currently is:" + brokerURI.getPath()).build();
+                }
+                if (!("tcp".equals(brokerURI.getScheme()) || "ssl".equals(brokerURI.getScheme()))) {
+                    return new ValidationResult.Builder().subject(subject).valid(false).explanation("only the 'tcp' and 'ssl' schemes are supported.").build();
+                }
+            } catch (URISyntaxException e) {
+                return new ValidationResult.Builder().subject(subject).valid(false).explanation("it is not valid URI syntax.").build();
+            }
+            return new ValidationResult.Builder().subject(subject).valid(true).build();
+        }
+    };
+
+    public static final Validator RetainValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            if("true".equalsIgnoreCase(input) || "false".equalsIgnoreCase(input)){
+                return new ValidationResult.Builder().subject(subject).valid(true).build();
+            } else{
+                return StandardValidators.createAttributeExpressionLanguageValidator(AttributeExpression.ResultType.BOOLEAN, false)
+                        .validate(subject, input, context);
+            }
+
+        }
+    };
+
+    public static final PropertyDescriptor PROP_BROKER_URI = new PropertyDescriptor.Builder()
+            .name("Broker URI")
+            .description("The URI to use to connect to the MQTT broker (e.g. tcp://localhost:1883)")
+            .required(true)
+            .addValidator(BrokerValidator)
+            .build();
+
+
+    public static final PropertyDescriptor PROP_CLIENTID = new PropertyDescriptor.Builder()
+            .name("Client ID")
+            .description("MQTT client ID to use")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_USERNAME = new PropertyDescriptor.Builder()
+            .name("Username")
+            .description("Username to use when connecting to the broker")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_PASSWORD = new PropertyDescriptor.Builder()
+            .name("Password")
+            .description("Password to use when connecting to the broker")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```

Comment:
I don't want to impose limitations on passwords (who knows what people make their password) so I figure NON_EMPTY is valid.